src/lib/wind_estimator/WindEstimator.cpp

```diff
@@ -0,0 +1,274 @@
+{
+	Vector2f velIvar_constrained = { max(0.01f, velIvar(0)), max(0.01f, velIvar(1)) };
+
+	if (!_initialised) {
+		// try to initialise
+		_initialised =	initialise(velI, velIvar_constrained, true_airspeed);
+		return;
+	}
+
+	// assign helper variables
+	const float v_n = velI(0);
+	const float v_e = velI(1);
+	const float v_d = velI(2);
+
+	const float k_tas = _state(tas);
+
+	// compute kalman gain K
+	const float HH0 = sqrtf(v_d * v_d + (v_e - w_e) * (v_e - w_e) + (v_n - w_n) * (v_n - w_n));
+	const float HH1 = k_tas / HH0;
+
+	Matrix<float, 1, 3> H_tas;
+	H_tas(0, 0) = HH1 * (-v_n + w_n);
+	H_tas(0, 1) = HH1 * (-v_e + w_e);
+	H_tas(0, 2) = HH0;
+
+	Matrix<float, 3, 1> K = _P * H_tas.transpose();
+
+	const Matrix<float, 1, 1> S = H_tas * _P * H_tas.transpose() + _tas_var;
+
+	K /= (S._data[0][0]);
+
+	// compute innovation
+	const float airspeed_pred = _state(tas) * sqrtf((v_n - _state(w_n)) * (v_n - _state(w_n)) + (v_e - _state(w_e)) *
+				    (v_e - _state(w_e)) + v_d * v_d);
+
+	_tas_innov = true_airspeed - airspeed_pred;
+
+	// innovation variance
+	_tas_innov_var = S._data[0][0];
+
+	// apply correction to state
+	_state(w_n) += _tas_innov * K(0, 0);
+	_state(w_e) += _tas_innov * K(1, 0);
+	_state(tas) += _tas_innov * K(2, 0);
+
+	// update covariance matrix
+	_P = _P - K * H_tas * _P;
+
+	run_sanity_checks();
+}
+
+void
+WindEstimator::fuse_beta(const Vector3f &velI, const Quatf &q_att)
+{
+	if (!_initialised) {return;}
+
+	const float v_n = velI(0);
+	const float v_e = velI(1);
+	const float v_d = velI(2);
+
+	// compute sideslip observation vector
+	float HB0 = 2.0f * q_att(0);
+	float HB1 = HB0 * q_att(3);
+	float HB2 = 2.0f * q_att(1);
+	float HB3 = HB2 * q_att(2);
+	float HB4 = v_e - w_e;
+	float HB5 = HB1 + HB3;
+	float HB6 = v_n - w_n;
+	float HB7 = q_att(0) * q_att(0);
+	float HB8 = q_att(3) * q_att(3);
+	float HB9 = HB7 - HB8;
+	float HB10 = q_att(1) * q_att(1);
+	float HB11 = q_att(2) * q_att(2);
+	float HB12 = HB10 - HB11;
+	float HB13 = HB12 + HB9;
+	float HB14 = HB13 * HB6 + HB4 * HB5 + v_d * (-HB0 * q_att(2) + HB2 * q_att(3));
+	float HB15 = 1.0f / HB14;
+	float HB16 = (HB4 * (-HB10 + HB11 + HB9) + HB6 * (-HB1 + HB3) + v_d * (HB0 * q_att(1) + 2.0f * q_att(2) * q_att(3))) /
+		     (HB14 * HB14);
+
+	Matrix<float, 1, 3> H_beta;
+	H_beta(0, 0) = HB13 * HB16 + HB15 * (HB1 - HB3);
+	H_beta(0, 1) = HB15 * (HB12 - HB7 + HB8) + HB16 * HB5;
+	H_beta(0, 2) = 0;
+
+	// compute kalman gain
+	Matrix<float, 3, 1> K = _P * H_beta.transpose();
+
+	const Matrix<float, 1, 1> S = H_beta * _P * H_beta.transpose() + _beta_var;
+
+	K /= (S._data[0][0]);
+
+	// compute predicted side slip angle
+	Vector3f rel_wind = Vector3f(velI(0) - _state(w_n), velI(1) - _state(w_e), velI(2));
+	Dcmf R_body_to_earth = Quatf(q_att);
+	rel_wind = R_body_to_earth.transpose() * rel_wind;
+
+	if (fabsf(rel_wind(0)) < 0.1f) {
+		return;
+	}
+
+	// use small angle approximation, sin(x) = x for small x
+	const float beta_pred = rel_wind(1) / rel_wind(0);
+
```

Comment:
hi @priseborough  Can i ask you that what `beta_pred = rel_wind(1) / rel_wind(0)` means? Means the a slope, tilt or somethings others?  Thank you.  And if it can means the slope, why?